### PR TITLE
More `vec_cast()` inheritance

### LINKED ISSUE #776: Can't cast <type> to <type>
``` r
library(vctrs)
TestS4 <- suppressWarnings(setClass("TestS4", contains = "integer"))
setMethod('[', 'TestS4', function(x, i, ...){ 
  TestS4(unclass(x)[i, ...])  
})
vec_c(TestS4(3:1))
#> Error: Can't cast <TestS4> to <TestS4>.
```

May be the same as https://github.com/tidyverse/glue/issues/166

```r
library(glue)
vec_c(glue(""))
#> Error: Can't cast <glue> to <glue>.
vec_c(glue(""))
#> Error: Can't cast <glue> to <glue>.
```

### PR DESCRIPTION
**Pull request by DavisVaughan**
Closes #776 with the usage of `is_same_type()` in `vec_default_cast()`

Addresses https://github.com/r-lib/vctrs/issues/753#issuecomment-576969414 by adding `is.object(to)` checks in the first level of casting dispatch.

@lionel- you specifically tested that something else was supposed to happen, i.e.

```r
expect_is(vec_cast(ptype, x), "foobar")
```

should this not have been an error? You can cast from, for example, `<double>` to `<foobar>` without writing a method for it? I found this super confusing when writing an S4 `bdays` class that "contained" an integer, and then was automatically able to do:

```r
> vec_cast(1, bdays(2, schedule()))
[1] 1
```

(That is the same problem as https://github.com/r-lib/vctrs/issues/753#issuecomment-576969414)

Also, part of the confusion came from the fact that a `vctrs_vctr` was being used in the tests, which has its own `vec_cast.vctrs_vctr.default()` method. I've changed the test to use a non-vctrs_vctr object

### CODE REVIEW

**Comment by romainfrancois** [comment]
Might not be the same issue, but I'm seeing this: 

``` r
library(vctrs)
library(glue)

vec_c("", glue(""), .ptype = character())
#> Error: Can't cast <glue> to <character>.
vec_c(glue(""), "", .ptype = character())
#> Error: Can't cast <glue> to <character>.
```

<sup>Created on 2020-01-22 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9000)</sup>

**Comment by lionel-** [comment]
> should this not have been an error?

yeah as discussed elsewhere we're in the process of preventing inheritance and restricting the cast methods. But we'll change that in 0.3.0. I didn't want to introduce too many breaking changes.

**Comment by lionel-** [comment]
@romainfrancois This is a trickier question. This is about default coercion methods for foreign  types that explicitly inherit from their base type. There are two aspects:

1. Direction of promotions with `vec_ptype2()`
2. Narrowing and widening casts with `vec_cast()`.

(1) Determines which type is richer. In most cases, the subclass is the richer type, and so `vec_ptype2(glue(), chr())` should normally return `glue()`. This indicates we should in fact allow inheritance by default, in the base type methods for ptype2.

(2) Should allow both kind of casts. When narrowing to the base class, we just strip the attributes. But what should we do when widening to the subclass? Copying over the attributes over would be the pragmatic solution. This makes the assumption that the attributes do not depend on the data, which works for simple classes like `glue()`. Here again, it seems that we should allow inheritance.

I think we lost track of the goal to be helpful with classes that inherit explicitly from their base type when we disallowed inheritance for base types to avoid creating new dispatch paths for `vctrs_vctr` classes. Given the considerations above, I now think we should set `new_vctr(inherit_base_type = )` to `FALSE` by default, instead of `TRUE`. This way users of `new_vctr()` have to be explicit about their typing hierarchy but we have an accommodating default for foreign classes.

**Comment by lionel-** [comment]
If we take this direction, there'll be a semantic difference between inheritance from a base type and inheritance from a normal class. The former will determine a default subtyping hierarchy, while the latter won't. This difference may introduce confusion, but the pay-off of an accommodating default for foreign classes might be worth it.

**Comment by lionel-** [comment]
On second thought we should probably disallow narrowing casts to the base class by default, as this seems safer.

@romainfrancois Did you have an important use case in mind for `vec_c("", glue(""), .ptype = character())`?

**Comment by DavisVaughan** [comment]
> disallow narrowing casts to the base class

Is this what we already do? We seem to disallow `<foo> -> <double>`, but allow `<double> -> <foo>` (but that doesn't return a `"foo"`, which is my main complaint).

``` r
library(vctrs)

x <- 1L
class(x) <- c("foo", "integer")

# widening cast - allowed but doesn't give a "foo"
vec_cast(1, x)
#> [1] 1

# narrowing cast - not allowed
vec_cast(x, 1)
#> Error: Can't cast `x` <foo> to `to` <double>.
```

<sup>Created on 2020-01-22 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9000)</sup>

**Comment by lionel-** [comment]
@DavisVaughan It seems that CRAN vctrs even has it backwards.

```r
# Widening
vec_cast(1L, x)
#> [1] 1

# Narrowing
vec_cast(x, 1L)
#> [1] 1
#> attr(,"class")
#> [1] "foo"     "integer"
```

Let's decide on the proper semantics for simple foreign classes like glue and implement them for 0.2.2.

**Comment by romainfrancois** [comment]
@lionel- I have the helper `dplyr:::glue_c()` but I can work around this, no problem.

File: R/type-bare.R [hunk 1]
```diff
@@ -617,7 +635,10 @@ vec_cast.character.default <- function(x, to, ..., x_arg = "x", to_arg = "to") {
 #' @export vec_cast.list
 #' @method vec_cast list
 #' @export
-vec_cast.list <- function(x, to, ...) {
+vec_cast.list <- function(x, to, ..., x_arg = "x", to_arg = "to") {
+  if (is.object(to)) {
```
Comment (krlmlr) [inline R/type-bare.R]: Do we need to check `vec_is(x)` here?

**Comment by lionel-** [comment]
Superseded by #963.

**Pull request closed without merging**